processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultFramedOnHeapAggregatable.java

```diff
@@ -85,10 +88,234 @@ public RowsAndColumns aggregateAll(
+          }
+          if (rangeId >= numRanges) {
+            return numRanges;
+          }
+          return rangeId;
+        }
+      };
+    }
+  }
+
+
+  /**
+   * Basic [a,b) interval; left inclusive/right exclusive.
+   */
+  static class Interval implements Iterable<Integer>
+  {
+    int a;
+    int b;
+
+    public static Interval of(int u, int v)
+    {
+      return new Interval(u, v);
+    }
+
+    public Interval(int u, int v)
+    {
+      this.a = u;
+      this.b = v;
+    }
+
+    @Override
+    public Iterator<Integer> iterator()
+    {
+      return new Iterator<Integer>()
+      {
+        int current = a;
+
+        @Override
+        public Integer next()
+        {
+          if (!hasNext()) {
+            throw new IllegalStateException();
+          }
+          return current++;
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+          return current < b;
+        }
+      };
+
+    }
+  }
+
+  /**
+   * Represents a range between U (inclusive) and V (exclusive).
+   */
+  static class Range
+  {
+    Interval outputRows;
+    Interval inputRows;
+
+    public Range(Interval outputRows, Interval inputRows)
+    {
+      this.outputRows = outputRows;
+      this.inputRows = inputRows;
+    }
+  }
+
+  static class AggCell
+  {
+    private AggregatorFactory[] aggFactories;
+    Interval currentRows = new Interval(0, 0);
+    private final AtomicInteger rowIdProvider;
+    private final ColumnSelectorFactory columnSelectorFactory;
+
+    private final Aggregator[] aggregators;
+
+
+    AggCell(AppendableRowsAndColumns rac, AggregatorFactory[] aggFactories)
+    {
+      this.aggFactories = aggFactories;
+      aggregators = new Aggregator[aggFactories.length];
+
+      rowIdProvider = new AtomicInteger(0);
+      columnSelectorFactory = ColumnSelectorFactoryMaker.fromRAC(rac).make(rowIdProvider);
+
+      newAggregators();
+    }
+
+    private void newAggregators()
+    {
+      for (int i = 0; i < aggFactories.length; i++) {
+        aggregators[i] = aggFactories[i].factorize(columnSelectorFactory);
+      }
+    }
+
+    /**
+     * Reposition aggregation window to reflect the given rows.
+     */
+    public void moveTo(Interval newRows)
+    {
+      // incremental addition of additional values
+      if (currentRows.a == newRows.a && currentRows.b < newRows.b) {
+        for (int i = currentRows.b; i < newRows.b; i++) {
+          aggregate(i);
+        }
+        currentRows = newRows;
+        return;
+      }
+      newAggregators();
+      for (int i : newRows) {
+        aggregate(i);
+      }
+      currentRows = newRows;
+    }
+
+    public void setOutputs(Object[][] results, Interval outputRows)
+    {
+      for (int aggIdx = 0; aggIdx < aggFactories.length; aggIdx++) {
+        Object aggValue = aggregators[aggIdx].get();
+        for (int rowIdx = outputRows.a; rowIdx < outputRows.b; rowIdx++) {
+          results[aggIdx][rowIdx] = aggValue;
+        }
```

Comment:
If, e.g., this method is called, then `aggregate()` is called before `newAggregators()` is called, it's totally acceptable (though not currently documented on the `Aggregator` interface) that the underlying object itself gets mutated as part of the subsequent `aggregate()` call.  For a cumulative aggregation, you need to do a `.get()` and then feed that as the first input back into a new Aggregator object.